lib/templatesimplifier.cpp

```diff
@@ -1068,7 +1068,7 @@ void TemplateSimplifier::useDefaultArgumentValues(TokenAndName &declaration)
                 do {
                     instantiationArgs[index].push_back(tok1);
                     tok1 = tok1->next();
-                } while (tok1 && tok1 != endLink);
+                } while (tok1 != endLink);
```

Comment:
if tok1->next() returns NULL then we should not continue looping right? Are you 110% sure that we will never reach the end of the tokenlist?

Reply:
If tok1 is NULL here, then we dereference it (calling `tok1->next()`) in the for loop iteration on line 1065 when tok1 in NULL).
Besides, I believe there's always a full chain from `tok1` to `tok1->link()` which is what we are following here. Unless of course tok1->link() is wrong, but it means there's an issue somewhere else (in which case this code is undefined behavior anyway because of the loop iteration).

Maybe a _safer_ approach would be to test for tok1 being NULL in the loop but I believe this would do more harm than good, as it should not happen.

Reply:
> If tok1 is NULL here, then we dereference it (calling tok1->next()) in the for loop iteration on line 1065 when tok1 in NULL).

No.. if tok1 is last token in tokenlist. then tok1->next() will return null. and then the loop should finish.